### PR TITLE
(refactor) Refactor test setup to leverage caching by `turbo`

### PR DESCRIPTION
**Pull request by denniskigen**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary

This PR refactors the test setup in this monorepo to leverage the caching capabilities of `turbo`. The current setup does properly leverage `turbo` for the following reasons:

- It uses a single Jest configuration, which gets used by the `test` scripts in the root-level `package.json` to run tests across the entire monorepo. As such, all the tests get re-run regardless of whether the associated files changed.

Ideally, we would like to leverage caching when executing tests. This PR introduces exactly that by moving testing-related scripts from the root-level `package.json` to each `package.json` file of each package. Additionally, this PR also:

- Moving mocks from the root-level `__mocks__` directory to package-specific `__mocks__` directories. This step also involves moving shared mocks to the `test-helpers` file.
- Cleaning up a bunch of test warnings and adding some useful refactors.

With these changes, we can now fully leverage caching when running tests.

## Video

https://github.com/openmrs/openmrs-esm-patient-management/assets/8509731/c2da7145-7e54-4cbe-9cf3-d03a97a7115a

## Related Issue
*None*

## Other
*None*

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -9,16 +12,17 @@ module.exports = {
   moduleNameMapper: {
     '\\.(s?css)$': 'identity-obj-proxy',
     '@openmrs/esm-framework': '@openmrs/esm-framework/mock',
-    '^lodash-es/(.*)$': 'lodash/$1',
-    '^uuid$': '<rootDir>/node_modules/uuid/dist/index.js',
     '^dexie$': require.resolve('dexie'),
+    '^lodash-es/(.*)$': 'lodash/$1',
+    '^react-i18next$': path.resolve(__dirname, 'react-i18next.js'),
+    '^uuid$': path.resolve(__dirname, 'node_modules', 'uuid', 'dist', 'index.js'),
```
Comment (vasharma05) [inline jest.config.js:18]: What does path.resolve do?

File: packages/esm-active-visits-app/package.json [hunk 2]
```diff
@@ -13,7 +13,10 @@
     "debug": "npm run serve",
     "build": "webpack --mode production",
     "analyze": "webpack --mode=production --env.analyze=true",
-    "lint": "eslint src --ext ts,tsx",
+    "lint": "cross-env TIMING=1 eslint src --ext ts,tsx",
+    "test": "cross-env TZ=UTC jest --config jest.config.js --verbose false --passWithNoTests",
```
Comment (vasharma05) [inline packages/esm-active-visits-app/package.json:17]: Can't we just use `../../jest.config.js`? This actually creates duplication and becomes tiring in a breaking change.
Thanks!

Comment (vasharma05) [reply packages/esm-active-visits-app/package.json:17]: I know we are using `require('../../jest.config.js')` but we can't we use it directly from the root?

File: packages/esm-appointments-app/src/admin/appointment-services/appointment-services.component.tsx [hunk 3]
```diff
@@ -57,94 +57,104 @@ const AppointmentServices: React.FC<AppointmentServicesProps> = () => {
...
-              pattern="([\d]+:[\d]{2})"
-              name="startTime"
-              value={props.values.startTime}
-              onChange={props.handleChange}
-              labelText={t('startTime', 'Start Time')}
-              id="start-time-picker">
-              <TimePickerSelect
-                name="startTimeTimeFormat"
+            <Layer>
+              <TextInput
```
Comment (vasharma05) [inline packages/esm-appointments-app/src/admin/appointment-services/appointment-services.component.tsx:61]: Is this relevant to this PR only?

**Review by vasharma05 — COMMENTED** [review]
Everything else looks good!
Just doubtful that adding Layer is okay to add in here.
Thanks!

File: packages/esm-appointments-app/src/admin/appointment-services/appointment-services.component.tsx (hunk 3 above)
Comment (denniskigen) [reply packages/esm-appointments-app/src/admin/appointment-services/appointment-services.component.tsx:61]: As I explained on the call, yes. This clears warnings about the soon-to-be-deprecated `light` prop for Carbon components when running tests.

File: turbo.json [hunk 4]
```diff
@@ -2,26 +2,13 @@
...
-    },
-    "dev": {
-      "cache": false
-    }
+    "coverage": {},
+    "extract-translations": {},
+    "lint": {},
+    "test": {},
+    "test:watch": {},
+    "typescript": {}
```
Comment (denniskigen) [inline turbo.json:12]: Looks like I wasn't using topological dependencies in pipelines correctly. I've amended these pipelines accordingly.

File: jest.config.js (hunk 1 above)
Comment (denniskigen) [reply jest.config.js]: Resolves the path relative to the current working directory. As discussed on the call, we shouldn't need this, and I've reverted back to using `<rootDir>`.

File: packages/esm-active-visits-app/package.json (hunk 2 above)
Comment (denniskigen) [reply packages/esm-active-visits-app/package.json]: Thanks for the spot. Makes sense!

File: turbo.json (hunk 4 above)
Comment (ibacher) [reply turbo.json:12]: I would've thought that toplogical dependencies would be required for typescript and _maybe_ linting (since the linter might also check types)?

**Review by ibacher — APPROVED** [review]

File: turbo.json (hunk 4 above)
Comment (denniskigen) [reply turbo.json:12]: Yeah, I went a bit overboard with that. I've amended things in the latest commit.

**Comment by ibacher** [comment]
This looks good to me, although, I probably look more later. I'm actually ok with having a `jest.config.js` for each project... just like we have a webpack.config.js, even though all that does is re-export a library. It may not be something we need now, but if an app develops the need for a specialised jest config, we'd already have the files.

**Comment by denniskigen** [comment]
The patient chart PR revealed an issue with using a single jest config at the root of the project and referencing it in all the child packages' test script as the argument to `--config` - performance slows down to a literal crawl. So, using multiple configs as is the case with the latest commits is probably the best approach.

File: packages/esm-active-visits-app/package.json (hunk 2 above)
Comment (denniskigen) [reply packages/esm-active-visits-app/package.json:17]: Update: Per our discussion from earlier today, we've chosen to go with multiple jest configs because of the performance issues when using a single jest config.
